### ISSUE TITLE
RPC prefetch is too aggressive for Insert queries

### ISSUE THREAD

**Issue by flyingsilverfin** [graknlabs/grakn#6072]
## Problem to Solve
1. When doing insert queries, or match-insert queries, we do not want to initiate precomputation (the number of prefetch we compute and send, based on the batch size and latency)
2. For all queries, we should not use the time delta we currently use (latency = time from user sending request, to first answer), as it can include either query planning time or the time to `match` all answers, then execute `insert` or `delete` on each of them. Please double check that our understanding is correct on this @alexjpwalker.

## Proposed Solution
1. for insert queries, we should initiate sending batches of answers (especially because it costs us nothing to actually compute them - they are already computed!) when the user requests the first batch, so in total there will be two round trips to initiate this. We can then continue with latency compensation and batching

2. For latency compensation, we should compensate the normal latency, not including query-planning, and/or the time to perform match-insert, or match-delete.

In essence, we want to not overly prefetch answer to the user, as apparently sending too many grpc messages has significant CPU time overhead. A clear win is to not pre-send any answers for insert queries, as outlined in 1. above.

**Comment by flyingsilverfin** [comment]
Here's the hack i used to test this with Grami:
```
        <T> void beginIteration(TransactionProto.Transaction.Req request, Iterator<T> iterator, int batchSize, Function<List<T>, TransactionProto.Transaction.Res> responseBuilderFn) {
            QueryProto.Query.Req req = request.getQueryReq();
            final String requestId = request.getId();
            BatchingIterator<T> batchingIterator;
            if (req.getReqCase() == INSERT_REQ) {
                batchingIterator = new BatchingIterator<>(requestId, iterator, responseBuilderFn, 1, 0);
            } else {
                final int latencyMillis = request.getLatencyMillis();
                batchingIterator = new BatchingIterator<>(requestId, iterator, responseBuilderFn, batchSize, latencyMillis);
            }
            iterators.compute(requestId, (key, oldValue) -> {
                if (oldValue == null) return batchingIterator;
                else throw GraknException.of(DUPLICATE_REQUEST, requestId);
            });
            batchingIterator.iterateBatch();
        }

```

probably worth reproducing we do make significant performance gains for scenario 1.

**Comment by alexjpwalker** [comment]
>     2. For all queries, we should not use the time delta we currently use (latency = time from user sending request, to first answer), as it can include either query planning time or the time to `match` all answers, then execute `insert` or `delete` on each of them. Please double check that our understanding is correct on this @alexjpwalker.

I don't think this is correct, @flyingsilverfin . Latency compensation continues to fetch answers for an additional N milliseconds, where N is the network latency between the client and the server.

Clients are responsible for computing and reporting their own network latency on a per-transaction basis. The computation is performed as follows:

- Start a timer T1
- Open a transaction by sending an `Open` request to the server
- When the server receives the request, _it_ starts a timer, T2
- When the server has opened the transaction and sends the `Open` response, it includes the elapsed time of T2, calling it the "computation time"
- When the client receives the response, it takes the elapsed time of T1, subtracts the _computation time_, and saves that as the network latency
- That latency is then sent together with each subsequent query in this transaction

On reviewing the server code with Haikal, we believe it is functioning as intended.

In conclusion I believe we should focus simply on fixing issue (1) and making Insert queries no longer prefetch any answers.

**Comment by flyingsilverfin** [comment]
alright, sounds good to me!
